src/drivers/amd/common/acp_dma.c

```diff
@@ -0,0 +1,396 @@
+	acp_dma_chan->config[channel->index].wr_size = 0;
+}
+
+/* Stop the requested channel */
+static int acp_dma_stop(struct dma_chan_data *channel)
+{
+	acp_dma_cntl_0_t	 dma_cntl;
+	acp_dma_ch_sts_t	 ch_sts;
+	uint32_t dmach_mask;
+	uint32_t delay_cnt = 10000;
+
+	switch (channel->status) {
+	case COMP_STATE_READY:
+	case COMP_STATE_PREPARE:
+		return 0; /* do not try to stop multiple times */
+	case COMP_STATE_PAUSED:
+	case COMP_STATE_ACTIVE:
+		break;
+	default:
+		return -EINVAL;
+	}
+	channel->status = COMP_STATE_READY;
+	dmach_mask = (1 << channel->index);
+	dma_cntl = (acp_dma_cntl_0_t)dma_chan_reg_read(channel, ACP_DMA_CNTL_0);
+	/* Do the HW stop of the DMA */
+	/* set DMAChRst bit to stop the transfer */
+	dma_cntl.bits.dmachrun      = 0;
+	dma_cntl.bits.dmachiocen    = 0;
+	dma_chan_reg_write(channel, ACP_DMA_CNTL_0, dma_cntl.u32all);
+	ch_sts = (acp_dma_ch_sts_t)dma_reg_read(channel->dma, ACP_DMA_CH_STS);
+	if (ch_sts.bits.dmachrunsts & dmach_mask) {
+		/* set the reset bit for this channel to stop the dma transfer */
+		dma_cntl.bits.dmachrst	 = 1;
+		dma_chan_reg_write(channel, ACP_DMA_CNTL_0, dma_cntl.u32all);
+	}
+
+	while (delay_cnt > 0) {
+		ch_sts = (acp_dma_ch_sts_t)dma_reg_read(channel->dma, ACP_DMA_CH_STS);
+		if (!(ch_sts.bits.dmachrunsts & dmach_mask)) {
+			/* clear the reset flag after successfully stopping the dma transfer
+			 * and break from the loop
+			 */
+			dma_cntl.bits.dmachrst = 0;
+			dma_chan_reg_write(channel, ACP_DMA_CNTL_0, dma_cntl.u32all);
+			break;
+		}
+		delay_cnt--;
+	}
+	return 0;
+}
+
+static int acp_dma_start(struct dma_chan_data *channel)
+{
+	acp_dma_cntl_0_t dma_cntl;
+	acp_dma_ch_sts_t dma_sts;
+	uint32_t chan_sts;
+	int ret = 0;
+	struct timer *timer = timer_get();
+	uint64_t deadline = platform_timer_get(timer) +
+				clock_ms_to_ticks(PLATFORM_DEFAULT_CLOCK, 1) * 500 / 1000;
+	if (channel->status != COMP_STATE_PREPARE &&
+	    channel->status != COMP_STATE_SUSPEND)
+		return -EINVAL;
+	channel->status = COMP_STATE_ACTIVE;
+	/* Clear DMAChRun before starting the DMA Ch */
+	dma_cntl =  (acp_dma_cntl_0_t)dma_chan_reg_read(channel, ACP_DMA_CNTL_0);
+	dma_cntl.bits.dmachrun = 0;
+	dma_cntl.bits.dmachiocen  = 0;
+	dma_chan_reg_write(channel, ACP_DMA_CNTL_0, dma_cntl.u32all);
+
+	dma_cntl =  (acp_dma_cntl_0_t)dma_chan_reg_read(channel, ACP_DMA_CNTL_0);
+	dma_cntl.bits.dmachrun	 = 1;
+	dma_cntl.bits.dmachiocen = 0;
+
+	/* set dmachrun bit to start the transfer */
+	dma_chan_reg_write(channel, ACP_DMA_CNTL_0, dma_cntl.u32all);
+
+	/* poll for the status bit
+	 * to finish the dma transfer
+	 * then initiate call back function
+	 */
+	dma_sts = (acp_dma_ch_sts_t)dma_reg_read(channel->dma, ACP_DMA_CH_STS);
+	chan_sts = dma_sts.u32all & (1 << channel->index);
+	while (chan_sts) {
+		if (deadline < platform_timer_get(timer)) {
+			dma_sts = (acp_dma_ch_sts_t)dma_reg_read(channel->dma, ACP_DMA_CH_STS);
+			chan_sts = dma_sts.u32all & (1 << channel->index);
```

Comment:
now you can move these 2 lines 174 and 175 above the `if` - between lines 172 and 173, because you execute them in both cases, but you also do the same in lines 170 and 171, so I'd propose to do
```
do {
	dma_sts = (acp_dma_ch_sts_t)dma_reg_read(channel->dma, ACP_DMA_CH_STS);
	chan_sts = dma_sts.u32all & (1 << channel->index);
	if (!chan_sts)
		return 0;
	/* consider adding a small delay here */
} while (platform_timer_get(timer) <= deadline);

tr_err(&acpdma_tr, "acp-dma: timed out for dma start");

return -ETIME;
```

Reply:
@saisurya-ch can we also address this? Repeating the two lines
```
	dma_sts = (acp_dma_ch_sts_t)dma_reg_read(channel->dma, ACP_DMA_CH_STS);
	chan_sts = dma_sts.u32all & (1 << channel->index);
```
3 times within a short code snippet where actually only a single repetition is needed is really too confusing.

Reply:
Done the changes.
Thanks @lyakh